sdk/samples/iot/hub/src/pnp_helper.h

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef _az_PNP_HELPER_H
+#define _az_PNP_HELPER_H
+
+#include <stdint.h>
+
+#include <azure/core/az_json.h>
+#include <azure/core/az_result.h>
+#include <azure/core/az_span.h>
+#include <azure/iot/az_iot_hub_client.h>
+
+#include "sample_pnp_thermostat_component.h"
+
+#define PNP_STATUS_SUCCESS 200
+#define PNP_STATUS_BAD_FORMAT 400
+#define PNP_STATUS_NOT_FOUND 404
+#define PNP_STATUS_INTERNAL_ERROR 500
+
+typedef void (*pnp_helper_property_callback)(
+    az_span component_name,
+    az_span property_name,
+    az_json_token* property_value,
+    int32_t version,
+    void* user_context_callback);
+
+/**
+ * @brief Callback which is called for each property found by the #pnp_helper_process_twin_data()
+ * API.
+ */
+typedef az_result (*pnp_append_property_callback)(az_json_writer* json_writer, void* context);
+
+/**
+ * @brief Gets the MQTT topic that must be used for device to cloud telemetry messages.
+ * @remark Telemetry MQTT Publish messages must have QoS At least once (1).
+ * @remark This topic can also be used to set the MQTT Will message in the Connect message.
```

Comment:
hub supports will messages?

Reply:
Apparently it does! Today I learned something :) 
See https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#using-the-mqtt-protocol-directly-as-a-device

> The device app can specify a Will message in the CONNECT packet. The device app should use `devices/{device_id}/messages/events/ or devices/{device_id}/messages/events/{property_bag}` as the Will topic name to define Will messages to be forwarded as a telemetry message. In this case, if the network connection is closed, but a DISCONNECT packet was not previously received from the device, then IoT Hub sends the Will message supplied in the CONNECT packet to the telemetry channel. The telemetry channel can be either the default Events endpoint or a custom endpoint defined by IoT Hub routing. The message has the `iothub-MessageType` property with a value of `Will` assigned to it.

Reply:
🎉